src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Grammar/GrammarGenerator.cs

```diff
@@ -0,0 +1,171 @@
+                field.Name == "Modifiers" ? RuleReference("Modifier") :
+                field.Name == "TextTokens" ? RuleReference(nameof(SyntaxKind.XmlTextLiteralToken)) : RuleReference(elementType))
+                    .Suffix(field.MinCount == 0 ? "*" : "+");
+
+        private static Production HandleTokenField(Field field)
+            => field.Kinds.Count == 0
+                ? HandleTokenName(field.Name)
+                : Join(" | ", field.Kinds.Select(k => HandleTokenName(k.Name))).Parenthesize(when: field.Kinds.Count >= 2);
+
+        private static Production HandleTokenName(string tokenName)
+            => GetSyntaxKind(tokenName) is var kind && kind == SyntaxKind.None ? RuleReference("SyntaxToken") :
+               SyntaxFacts.GetText(kind) is var text && text != "" ? new Production(text == "'" ? "'\\''" : $"'{text}'") :
+               tokenName.StartsWith("EndOf") ? new Production("") :
+               tokenName.StartsWith("Omitted") ? new Production("/* epsilon */") : RuleReference(tokenName);
+
+        private static SyntaxKind GetSyntaxKind(string name)
+            => GetMembers<SyntaxKind>().Where(k => k.ToString() == name).SingleOrDefault();
+
+        private static IEnumerable<TEnum> GetMembers<TEnum>() where TEnum : struct, Enum
+            => (IEnumerable<TEnum>)Enum.GetValues(typeof(TEnum));
+
+        private static Production RuleReference(string name)
+            => new Production(
+                s_normalizationRegex.Replace(name.EndsWith("Syntax") ? name[..^"Syntax".Length] : name, "_").ToLower(),
+                ImmutableArray.Create(name));
+
+        // Converts a PascalCased name into snake_cased name.
+        private static readonly Regex s_normalizationRegex = new Regex(
+            "(?<=[A-Z])(?=[A-Z][a-z]) | (?<=[^A-Z])(?=[A-Z]) | (?<=[A-Za-z])(?=[^A-Za-z])",
+            RegexOptions.IgnorePatternWhitespace | RegexOptions.Compiled);
+    }
+
+    internal struct Production : IComparable<Production>
+    {
+        public readonly string Text;
+        public readonly ImmutableArray<string> ReferencedRules;
+
+        public Production(string text, IEnumerable<string> referencedRules = null)
+        {
+            Text = text;
+            ReferencedRules = referencedRules?.ToImmutableArray() ?? ImmutableArray<string>.Empty;
+        }
+
+        public override string ToString() => Text;
+        public int CompareTo(Production other) => StringComparer.Ordinal.Compare(this.Text, other.Text);
+        public Production Prefix(string prefix) => new Production(prefix + this, ReferencedRules);
+        public Production Suffix(string suffix, bool when = true) => when ? new Production(this + suffix, ReferencedRules) : this;
+        public Production Parenthesize(bool when = true) => when ? Prefix("(").Suffix(")") : this;
+    }
+}
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static class GreenNode
+    {
+        internal const int ListKind = 1; // See SyntaxKind.
```

Comment:
Needed because `SyntaxKind` references `GreenNode.ListKind`.  Would be happy to just update `SYntaxKind` to hardcode this value.  It can't actually change, so i don't see the harm in that myself.